src/main/java/uk/co/qualitycode/utils/functional/Functional.java

```diff
@@ -1576,6 +1576,9 @@ public static <T> T last(final Iterable<T> input) {
         T state = null;
         for (final T element : input) state = element;
 
+        if (state == null)
+            throw new IllegalArgumentException("last(Iterable<T>): input must not be empty");
```

Comment:
ok, but it's not so much that the input is empty but that the input contains nulls, which is shocking.

Reply:
There probably ought to be a null check on input as well, though :-)

Reply:
I would have liked to check that it isn't empty upfront but its probably more efficient to check that the end state isn't null.

Reply:
input has an existing notNull check on it :)